### PR TITLE
Setup Pyodide with nox for easier local testing

### PR DESCRIPTION
**Pull request by messense**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -151,47 +151,35 @@ jobs:
...
-      - run: pip install nox
-      - uses: actions/cache@v3
-        with:
-          path: |
-            tests/pyodide
-          key: ${{ hashFiles('tests/*.js') }} - ${{ hashFiles('noxfile.py') }} - ${{ steps.setup-python.outputs.python-path }}
-      - uses: Swatinem/rust-cache@v1
       - name: Run tests
         run: |
           export PATH=$ORIG_PATH:$PATH
```
Comment (messense) [inline .github/workflows/test.yml:185]: @hoodmane Could explain why is this needed? Otherwise it throws the following error:

```
warning: no blob constructor, cannot create blobs with mimetypes
warning: no BlobBuilder
/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:5716
          throw ex;
          ^
TypeError: wasm function signature contains illegal type
    at lseek (<anonymous>:wasm-function[11109]:0x55491b)
    at <anonymous>:wasm-function[6841]:0x34ecde
    at method_call_trampoline (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:8785:33)
    at <anonymous>:wasm-function[1078]:0x137dc0
    at PyObject_VectorcallMethod (<anonymous>:wasm-function[954]:0x13081e)
    at <anonymous>:wasm-function[6926]:0x353e00
    at <anonymous>:wasm-function[6[89](https://github.com/PyO3/maturin/runs/7051097339?check_suite_focus=true#step:10:90)1]:0x35124e
    at <anonymous>:wasm-function[68[90](https://github.com/PyO3/maturin/runs/7051097339?check_suite_focus=true#step:10:91)]:0x35114d
    at <anonymous>:wasm-function[2222]:0x17fe6f
    at method_call_trampoline (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:8785:33)
```

File: noxfile.py [hunk 2]
```diff
@@ -1,12 +1,59 @@
...
+def setup_pyodide(session: nox.Session):
+    tests_dir = Path("./tests").resolve()
+    with session.chdir(tests_dir):
+        session.run(
+            "npm",
+            "i",
+            "--no-save",
+            f"pyodide@{PYODIDE_VERSION}",
+            "prettier",
+            "node-fetch",
```
Comment (messense) [inline noxfile.py:32]: Install node-fetch does not help

https://github.com/PyO3/maturin/runs/7051143312?check_suite_focus=true

```
PythonError: Traceback (most recent call last):
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 234, in __step
    result = coro.throw(exc)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 505, in install
    await transaction.gather_requirements(requirements)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 260, in gather_requirements
    await gather(*requirement_promises)
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 234, in __step
    result = coro.throw(exc)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 274, in add_requirement
    await self.add_wheel(wheel, extras=set())
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 391, in add_wheel
    await self.gather_requirements(wheel.requires(extras))
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 260, in gather_requirements
    await gather(*requirement_promises)
  File "/lib/python3.10/asyncio/futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/lib/python3.10/asyncio/tasks.py", line 232, in __step
    result = coro.send(None)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 264, in add_requirement
    return await self.add_requirement_inner(req)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 360, in add_requirement_inner
    metadata = await _get_pypi_json(req.name, self.fetch_kwargs)
  File "/lib/python3.10/site-packages/micropip/_micropip.py", line 41, in _get_pypi_json
    metadata = await fetch_string(url, fetch_kwargs)
  File "/lib/python3.10/site-packages/micropip/_compat_in_pyodide.py", line 33, in fetch_string
    return await (await pyfetch(url, **kwargs)).string()
  File "/lib/python3.10/site-packages/pyodide/http.py", line [228](https://github.com/PyO3/maturin/runs/7051143312?check_suite_focus=true#step:11:229), in pyfetch
    from js import fetch as _jsfetch
ImportError: cannot import name 'fetch' from 'js' (unknown location)
    at new_error (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:8790:33)
    at wasm://wasm/0[244](https://github.com/PyO3/maturin/runs/7051143312?check_suite_focus=true#step:11:245)357a:wasm-function[272]:0xe8b11
    at wasm://wasm/0244357a:wasm-function[371]:0xecd12
    at method_call_trampoline (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:8785:33)
    at wasm://wasm/0244357a:wasm-function[935]:0x12f804
    at wasm://wasm/0244357a:wasm-function[3185]:0x200362
    at wasm://wasm/0244357a:wasm-function[1848]:0x16aee8
    at wasm://wasm/0244357a:wasm-function[940]:0x12fd14
    at wasm://wasm/0244357a:wasm-function[941]:0x12fe0e
    at wasm://wasm/0244357a:wasm-function[942]:0x12feb1
    at wasm://wasm/0244357a:wasm-function[3011]:0x1e9acb
    at wasm://wasm/0244357a:wasm-function[3008]:0x1e3ade
    at wasm://wasm/0244357a:wasm-function[943]:0x12fef4
    at wasm://wasm/0244357a:wasm-function[3020]:0x1eb8c0
    at wasm://wasm/0244357a:wasm-function[3011]:0x1e950d
    at wasm://wasm/0244357a:wasm-function[3008]:0x1e3ade
    at wasm://wasm/0244357a:wasm-function[943]:0x12fef4
    at wasm://wasm/0244357a:wasm-function[366]:0xec9de
    at Object.Module.callPyObjectKwargs (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:4066:31)
    at Object.Module.callPyObject (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:4089:25)
    at Timeout.wrapper [as _onTimeout] (/home/runner/work/maturin/maturin/tests/node_modules/pyodide/pyodide.asm.js:7042:27)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7) {
  __error_address: 17766960
}
```

**Comment by messense** [comment]
Everything works fine locally, odd.

File: .github/workflows/test.yml (hunk 1 above)
Comment (messense) [reply .github/workflows/test.yml]: OK, I think it's `setup-emsdk` added an old version of Node to `PATH`.

Comment (hoodmane) [reply .github/workflows/test.yml]: Sorry I didn't leave a comment about that could have saved you the time working this out.

File: noxfile.py [hunk 3]
```diff
@@ -1,12 +1,58 @@
...
+                "../prettier/bin-prettier.js",
+                "-w",
+                "pyodide.asm.js",
+                external=True,
+            )
+            emscripten_version = (
+                subprocess.check_output(
+                    '''node -p "require('./repodata.json').info.platform.split('_').slice(1).join('.')"''',
+                    shell=True,
+                )
```
Comment (hoodmane) [inline noxfile.py:46]: Would be simpler to use `json.load` here.

File: noxfile.py [hunk 4]
```diff
@@ -1,12 +1,58 @@
...
+
+@nox.session(name="setup-pyodide")
+def setup_pyodide(session: nox.Session):
+    tests_dir = Path("./tests").resolve()
+    with session.chdir(tests_dir):
+        session.run(
+            "npm",
+            "i",
+            "--no-save",
+            f"pyodide@{PYODIDE_VERSION}",
```
Comment (hoodmane) [inline noxfile.py:30]: Maybe pass the Pyodide version as an argument to the session rather than as an environment variable for easier command line use?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -151,51 +151,39 @@ jobs:
   test-emscripten:
     name: Test Emscripten
     runs-on: ubuntu-latest
+    env:
+      PYODIDE_VERSION: '0.21.0-alpha.2'
+      PYTHON_VERSION: '3.10.2'
```
Comment (hoodmane) [inline .github/workflows/test.yml:156]: After https://github.com/pyodide/pyodide/pull/2786 we should be able to calculate `PYTHON_VERSION` from `repodata.json` the same way as `EMSCRIPTEN_VERSION`.
